Runner cleanup

j-tobler opened the pull request:
Some initial progress towards cleaning up RunUtils.

Created classes Transform, StaticAnalysis and AnalysisManager. Refactored doCleanup, prepareForTranslation, generateRelyGuaranteeConditions, and generateFunctionSummaries to use the Transform class.

ailrst commented:
`./mill scalafmt.reformat` will run the autoformatter

ailrst commented on src/main/scala/analysis/AnalysisManager.scala:
```diff
@@ -0,0 +1,69 @@
+//   val ssaResults = register(/* todo */)
+//   val graResult = register(/* todo */)
+//   val intraDomain = register(/* todo */)
+//   val interDomain = register(/* todo */)
```
Like I said, these analyses/transforms are all super tightly coupled so we don't want to break them down in the pass manager.

katrinafyi commented on src/main/scala/analysis/AnalysisManager.scala:
```diff
@@ -0,0 +1,59 @@
+    // allows this memoizer to be called like a function
+    def apply(): ReturnType = {
+      // pass this analysis manager and its associated program to the static analysis
+      if memo.isEmpty then memo = Some(analysis(AnalysisManager.this))
+      memo.get
```
use pattern match

katrinafyi commented on src/main/scala/ir/transforms/StripUnreachableFunctions.scala:
```diff
@@ -47,5 +50,34 @@ def stripUnreachableFunctions(p: Program, depth: Int = Int.MaxValue): Unit = {
-
 }
+
+def getStripUnreachableFunctionsTransform(depth: Int): Transform =
+  SingleTransform(
```
i dont really like the need to wrap parametrised transforms in a get method. maybe the Transform class can be changed to take an additional custom parameter.

katrinafyi commented on src/main/scala/ir/transforms/Transform.scala:
```diff
@@ -0,0 +1,113 @@
+import java.io.File
+
+// TransformConfig accepts Log instances which specify what kind of logs to dump for particular transforms
+trait Log {
+  def dump(ctx: IRContext, transformName: String): Unit
+}
+
+// dumps a blockgraph log
+case class BlockgraphLog(filenamePrefix: String) extends Log {
+  def dump(ctx: IRContext, transformName: String): Unit =
+    DebugDumpIRLogger.writeToFile(
+      File(s"${filenamePrefix}_blockgraph-${transformName}.dot"),
+      dotBlockGraph(ctx.program.mainProcedure)
+    )
+}
+
+// dumps an IR log
+case class IrLog(filenamePrefix: String) extends Log {
+  def dump(ctx: IRContext, transformName: String): Unit =
+    DebugDumpIRLogger.writeToFile(File(s"${filenamePrefix}_il-${transformName}.il"), pp_prog(ctx.program))
+}
```
what's the point of this? these could just be methods surely

katrinafyi commented on src/main/scala/ir/transforms/Transform.scala:
```diff
@@ -0,0 +1,113 @@
+  // optional code to run *after* performance has been measured but *before* any logs are dumped, e.g. post-run checks
+  protected def postRun(ctx: IRContext): Unit
+
+  // executes the transform with any modifications or book-keeping specified by the given config
+  def apply(ctx: IRContext, man: AnalysisManager, config: TransformConfig = emptyConfig): Unit = {
+    if (config.disabled.contains(this)) return
+    if (notice != "") then Logger.info(s"[!] ${notice}")
+    if (man.program ne ctx.program) {
+      // the program we are transforming should be the same one for which the analysis results were produced
+      throw new RuntimeException(
+        s"Transform '$name' was passed an AnalysisManager of an IR Program with a different " +
+          s"reference value than the program being transformed."
+      )
+    }
+    val maybeLogs = config.dumpLogs.get(this)
+    maybeLogs.foreach(_.foreach(_.dump(ctx, s"before-$name")))
+    timer.checkPoint("start")
+    val invalidation = transform(ctx, man, config) // run the actual transform and get the analysis results to clobber
+    timer.checkPoint("end")
+    postRun(ctx) // runs after performance checks, and before logging
+    maybeLogs.foreach(_.foreach(_.dump(ctx, s"after-$name")))
+    man.invalidate(invalidation) // clobber the specified analysis results
+  }
```
i think this method belongs more in the AnalysisManager, but shrug

katrinafyi commented on src/main/scala/ir/transforms/Transform.scala:
```diff
@@ -0,0 +1,113 @@
+
+/** A standard transform. Accepts an implementation function which modifies the given IR context in-place.
+  */
+case class SingleTransform(
+  name: String,
+  implementation: (ctx: IRContext, man: AnalysisManager) => man.Invalidation,
+  notice: String = ""
+) extends Transform {
```
single transform could be done as a special case of transform batch

katrinafyi reviewed (commented):
seems fine, it's very good to unify the dumping and logging things.

comments are more for future work

j-tobler commented:
This most recent merge had some pretty significant [conflicts](https://github.com/UQ-PAC/BASIL/pull/449/commits/33e1b96658c743f9e920ff2d559b4e15cc9e80eb).

Fortunately, most of these were due to @katrinafyi's [changes](https://github.com/UQ-PAC/BASIL/pull/477) to the IRLoading object, which @ailrst moved to a new [file](https://github.com/UQ-PAC/BASIL/pull/449/commits/33e1b96658c743f9e920ff2d559b4e15cc9e80eb#diff-43eb5437ef221f12226741beb6c83721136650f43eb95778358b4a26897757d6) in the feature branch. I resolved these by overwriting the new file with Kait's changes.

[Another](https://github.com/UQ-PAC/BASIL/pull/449/commits/33e1b96658c743f9e920ff2d559b4e15cc9e80eb#diff-a4596d2dee101a61805900eaaf56e60b6ad4ec25c62fd85b5428215e53e2ab22) was due to the new loop invariant generation pass, which I resolved by creating a new reducibleLoops function in AnalysePipelineMRA.scala, effectively re-creating the function added to the now-removed StaticAnalysis object in RunUtils.

One important change that I'm not sure about is the tests. Particularly, I removed some tests from [IntervalDSATest.scala](https://github.com/UQ-PAC/BASIL/pull/449/commits/33e1b96658c743f9e920ff2d559b4e15cc9e80eb#diff-536c818c47415d320b440436e3a7323c2cf9e3fec3f3cf55d33e2e3b97cdfbde) in an attempt to capture the latest changes made to the file. I'm not sure if this was correct and it would be good to check this with the author of the tests.

To ensure the refactor preserves the old behaviour, I loosely combed through the prepareForTranslation and doCleanup transforms and didn't find any discrepancies with the original functions in RunUtils. The doSimplify transform is still under development in Simp.scala.
